Use the rbd showmapped command after mapping, as rbd outputs nothing.

slepp opened the pull request:
On the distribution of Ceph I'm using, rbd outputs nothing when it maps. Using the showmapped command after the map returns gives the JSON result, which is iterated for the new device name.

mapuri commented on storage/backend/ceph/internals.go:
```diff
@@ -6,14 +6,30 @@ import (
 func (c *Driver) mapImage(do storage.DriverOptions) (string, error) {
-	blkdev, err := exec.Command("rbd", "map", do.Volume.Name, "--pool", do.Volume.Params["pool"]).Output()
-	device := strings.TrimSpace(string(blkdev))
+	var outputdata map[string]interface{}
```
Instead of a `map[string]interface{}`, I think it might be better to define `outputdata` to be a struct, this will avoid the dictionary lookup and type assertion below at line 29 and hence a chance of panic.

I am not sure what the json output looks like, but assuming it is a array of volume infos, we might be able to do something like this:

```
var outputData struct {
    volumes []struct {
        Name string `json:"name"`
    } `json:"volumes"`
}

json.Unmarshal(output, &outputData)
...
...

if outputdata[i].Name != do.Volume.Name {
...
}
```

mapuri commented on storage/backend/ceph/internals.go:
```diff
@@ -6,14 +6,30 @@ import (
+	var outputdata map[string]interface{}
+	var device string
+
+	_, err := exec.Command("rbd", "map", do.Volume.Name, "--pool", do.Volume.Params["pool"]).Output()
```
I think we should check for error here and not proceed if the command failed.

mapuri commented on storage/backend/ceph/internals.go:
```diff
@@ -6,14 +6,30 @@ import (
+	_, err := exec.Command("rbd", "map", do.Volume.Name, "--pool", do.Volume.Params["pool"]).Output()
+	output, err := exec.Command("rbd", "showmapped", "--format", "json").Output()
+
+	json.Unmarshal(output, &outputdata)
```
we should check for unmarshal errors here

mapuri commented on storage/backend/ceph/internals.go:
```diff
@@ -6,14 +6,30 @@ import (
+	output, err := exec.Command("rbd", "showmapped", "--format", "json").Output()
+
+	json.Unmarshal(output, &outputdata)
+	fmt.Println(outputdata)
```
we should log this using `log.Debugf()` function instead

mapuri commented on storage/backend/ceph/internals.go:
```diff
@@ -6,14 +6,30 @@ import (
+	device = ""
+
+	for i := range outputdata {
+		if outputdata[i].(map[string]interface{})["name"].(string) == do.Volume.Name {
```
if we are not able to define a struct for outputdata as suggested above we should check for key lookups and type assertion errors to avoid a panic.

mapuri commented on storage/backend/ceph/internals.go:
```diff
@@ -6,14 +6,30 @@ import (
+	for i := range outputdata {
+		if outputdata[i].(map[string]interface{})["name"].(string) == do.Volume.Name {
+			device = outputdata[i].(map[string]interface{})["device"].(string)
+		}
```
we can `break` out of the loop once we have found the matching volume.

mapuri commented:
@slepp thanks for the patch. I am not much familiar with ceph command line so will let @erikh comment on the approach as well. Other than that I have a few minor comments.

slepp commented:
Thanks for all the feedback. It's been about three years since I last wrote anything in Go, so it's a bit rusty. I'll review and make those changes in the next day or two. The struct makes sense and seems to be very constant and simple in format, so should do well in this context. I'll also add a check if rbd outputs the volume name as the original code used, for compatibility.

erikh commented:
@slepp for my curiosity, which version of ceph are you running?

erikh commented:
@slepp If you need help with Go, email me at the address on my github profile and we can work out a way to discuss this in a faster timeframe.

erikh commented on storage/backend/ceph/internals.go:
```diff
@@ -6,14 +6,30 @@ import (
+	json.Unmarshal(output, &outputdata)
+	fmt.Println(outputdata)
+
+	device = ""
```
This line is not necessary because it will be the default value of the declared `device` above.

slepp commented:
@erikh I'm using Ceph version 0.80.11 on the server running rbd.

An example JSON output is:

{"0":{"pool":"rbd","name":"test1","snap":"-","device":"\/dev\/rbd0"}}

I've updated things now with the suggestions you and @mapuri made.

erikh commented:
Ok, will review later today. Thank you so much!

erikh commented:
I'm sorry to ask this; if you could do this with your branch to kick the CI service:

```
git commit -a -s --amend
git push -f <my remote> <my branch>
```

The ansible used to set up the boxes occasionally fails due to some proxy issues where our Jenkins service lives. We're working on it. :)

slepp commented:
@erikh CI service kicked (hopefully :).

erikh commented:
Ok, sorry to do this after all the CI stuff as I didn't have time to look it over earlier.

Please test each error as you are, but when you get an error please return it, like so:

```go
return "", err
```

Otherwise this patch LGTM. @mapuri care to comment?

slepp commented:
Updated with error returns, including if the mapping isn't found (it would've cascaded out silently before).

Not sure if CI needs a kick or not.

mapuri commented on storage/backend/ceph/internals.go:
```diff
@@ -6,14 +6,46 @@ import (
+	if device == "" && err == nil {
+		output, err := exec.Command("rbd", "showmapped", "--format", "json").Output()
+
+		if err == nil {
```
this is just a minor code styling nit, by doing a `err != nil` check earlier, the large part of the code within the `if` can be kept indented to left, which is also easier t read.

```
if err != nil {
    return "", err
}

... the main logic left indented goes here
...
```

mapuri commented on storage/backend/ceph/internals.go:
```diff
@@ -6,14 +6,46 @@ import (
+
+		if err == nil {
+			err := json.Unmarshal(output, &rbdMaps)
+			if err == nil {
```
similar to comment above, if you check for `err != nil` and return the else block can be removed and the looping code will be indented to left making it more readable and less nested.

mapuri commented:
@slepp thanks for taking care of my comments. The code changes look good to me. I have a few minor comments around code styling, please see if you can incorporate those. Thanks again for your time to address comments!

slepp commented:
Updated the code there for readability (it was a bit confusing before). Moved the err check for the initial map command back near the top where it should actually be checked. Final return changed to always be device, nil, as we can't end up here with err != nil.

mapuri commented:
thanks @slepp! changes LGTM

erikh commented:
LGTM

erikh commented:
Thanks for your contribution!
